README.md
[![main](https://github.com/lorenzvanherwaarden/datetime-web-component/actions/workflows/main.yml/badge.svg)](https://github.com/lorenzvanherwaarden/datetime-web-component/actions/workflows/main.yml)
[![tests](https://github.com/lorenzvanherwaarden/datetime-web-component/actions/workflows/playwright.yml/badge.svg)](https://github.com/lorenzvanherwaarden/datetime-web-component/actions/workflows/playwright.yml)
![latest release](https://badgen.net/github/release/lorenzvanherwaarden/datetime-web-component)
![typescript](https://badgen.net/badge/icon/typescript?icon=typescript&label)

# Datetime web component

Lightweight and easily styleable datetime picker web component.

<p>
<img alt="Example of datetime-web-component" src="/public/default-style-light.png?raw=true" width="301">
<img alt="Example of datetime-web-component" src="/public/default-style-dark.png?raw=true" width="301">
</p>
Default UI in light & dark mode.

## Features

- Native web component which can be used anywhere; framework agnostic
- Style easily and extensively by setting CSS custom properties
- Lightweight; depends on only 1 dependency [floating-ui/dom](https://www.npmjs.com/package/@floating-ui/dom) to position dropdown
- Very performant
- Options to show/hide time and show/hide seconds
- Supports dark mode
- Subtle default styling

## Installation

```bash
npm install --save datetime-web-component
# yarn add datetime-web-component
```

## Usage

### 1. Import

```js
import 'datetime-web-component'
```

### 2. Style (optional)

Optionally style the datetime picker by defining CSS custom properties on the `datetime-web-component` tag. The list of all available custom properties to style is [available below](#styling):

```css
/* example styling */
datetime-web-component {
  --text-color: #353535;
  --font-family: 'Karla', sans-serif;
  --cell-width: 32px;
  --focus-color: #00aa44;
}
```

### 3. Include component and listen to changes

```html
<template>
  <datetime-web-component
    id="datetime"
    value="2022-12-07T12:12:03.000Z"
    show-seconds
  >
    <input slot="input" type="text" value="2022-12-07T12:12:03.000Z" />
  </datetime-web-component>
</template>
```

```js
const datetimeEl = document.getElementById('datetime')

// Listen to input event of the datetime-web-component for changes
datetimeEl.addEventListener('input', (event) => {
  inputEl.value = event.value
})
```

## API

### Properties

All properties, except the `isDayBlocked` (which isn't a primitive type), are exposed as property and attribute. This follows the web component attribute & properties [best practices](https://web.dev/custom-elements-best-practices/#attributes-and-properties).

| prop           | type                    | attribute      | description                                                                                                                                                                                                                                                                                                   | default       | required |
| -------------- | ----------------------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------- | -------- |
| `value`        | string \| null          | "value"        | String value of the date, like you would pass to the [Date constructor](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date) and should conform to the ISO 8601 format (YYYY-MM-DDTHH:mm:ss.sssZ). If no value is given, the datetime will be set to the current date. | `undefined`   | ✕        |
| `onlyDate`     | boolean                 | "only-date"    | Whether to hide the time picker.                                                                                                                                                                                                                                                                              | `false`       | ✕        |
| `showSeconds`  | boolean                 | "show-seconds" | Whether to show seconds in the time picker. This property has no effect when `onlyDate` is true.                                                                                                                                                                                                              | `false`       | ✕        |
| `isDayBlocked` | (date: Date) => boolean | ✕              | Days can be blocked by providing a function.                                                                                                                                                                                                                                                                  | `() => false` | ✕        |

### Events

You can listen to changes by [adding an event listener](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) with the type equal to `"input"` on the `<datetime-web-component />` element. The event contains a `value` property, which represents the new datetime in ISO 8601 format. This can be passed to the datetime-web-component directly as valid datetime. The event also contains a `date` property, which represents the `Date` representation of that new datetime. It is not necessary to feed the received `value` into the datetime-web-component again, as the internal value auto-updates when selecting a new datetime. It can be useful to update the input element's value with this event.

## Styling

You can style the datetime picker by overriding the CSS custom properties that are used in our css. These are all the exposed properties:

```css
datetime-web-component {
  --font-family: sans-serif;
  --font-size: 14px;
  --text-color: #222222;
  --muted-text-color: #777777;
  --disabled-text-color: #aaaaaa;
  --focus-color: #0066ff;

  /* dropdown */
  --background: #fafafa;
  --box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.08);
  --border-color: #dddddd;
  --border-width: 1px;
  --border-radius: 4px;
  --padding: 16px;

  /* date cell */
  --cell-width: 28px;
  --cell-height: 26px;
  --cell-border-radius: 6px;
  --cell-vertical-margin: 4px;
  --cell-horizontal-margin: 2px;
  --selected-cell-background: #e7e7e7;
  --selected-cell-font-weight: 700;
  --hover-cell-background: #efefef;

  /* time & year input */
  --input-font-size: 18px;
  --input-font-weight: 500;
  --input-border-radius: 4px;
  --input-hover-background: #eeeeee;
}
```

Dark mode styling can be done by using a media query

```css
@media (prefers-color-scheme: dark) {
  datetime-web-component {
    --text-color: #ffffff;
    --background: #333333;
  }
}
```

## Using datetime-web-component from CDN

You can use the datetime-web-component from a CDN via a script tag without a build step:

```html
<script src="https://unpkg.com/datetime-web-component@latest/dist/datetime-web-component.umd.js"></script>
```


CHANGELOG.md


## [0.5.0](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.4.0...v0.5.0) (2023-01-15)


### Features

* open datetime-web-component with focus ([#25](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/25)) ([9e3d24a](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/9e3d24a40bc7e8ccea194160a16aa86905f1fb51))
* use buttons for clickable cells ([#26](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/26)) ([875479e](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/875479e4347db992db3254b08bdfd80c74a170db))

## [0.4.0](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.3.0...v0.4.0) (2023-01-13)


### Features

* pass input as slot ([#21](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/21)) ([ca92619](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/ca92619f44ab8e8e99934f9e8d41743aa8c8e950))

## [0.3.0](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.2.0...v0.3.0) (2023-01-11)


### Features

* add vue example ([#14](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/14)) ([792a410](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/792a41074e54b4f283e7394aa47ea31de773fab4))


### Bug Fixes

* allow setting refId by property too ([#15](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/15)) ([a180a9f](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/a180a9f6c33a6555a8f5d284351009721670b638))

## [0.2.0](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.9...v0.2.0) (2023-01-06)

### Features

- allow to block certain dates ([#11](https://github.com/lorenzvanherwaarden/datetime-web-component/issues/11)) ([f4c98f4](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/f4c98f468d3d3879c2068fe7657fa928d0e1c832))

### [0.1.9](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.7...v0.1.9) (2023-01-05)

### [0.1.7](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.6...v0.1.7) (2023-01-05)

### [0.1.6](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.5...v0.1.6) (2023-01-05)

### [0.1.5](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.4...v0.1.5) (2022-08-21)

### [0.1.4](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.3...v0.1.4) (2022-08-17)

### Bug Fixes

- use correct attribute naming ([f64cc6e](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/f64cc6e82e757bba38f1391cacf14a01c368b4cf))

### [0.1.3](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.2...v0.1.3) (2022-08-15)

### [0.1.2](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.1...v0.1.2) (2022-08-15)

### [0.1.1](https://github.com/lorenzvanherwaarden/datetime-web-component/compare/v0.1.0...v0.1.1) (2022-08-15)

## 0.1.0 (2022-08-15)

### Features

- add dark mode colors ([d8171a2](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/d8171a259f78e9e5c603dd4b426e4b9aa10b5663))
- add month and year header ([bff048f](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/bff048f2dacdb125413a39d0a4223a9cfd49f47a))
- add more style variables ([26d6d30](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/26d6d30fdc314b33c365320081d6d9da0a0530ad))
- add year select view ([57be192](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/57be192a3a1d04af6bd9c0f3164cbfdbbee2a403))
- allow time input ([131ffce](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/131ffce4f682c81a14b6e66627008671986581e8))
- create cells and dispatch events ([5956037](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/5956037a4468ac3a8ad9bd6d5abe8d96d5211679))
- setup click listeners and first property handles ([7171f39](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/7171f394ad7e0ebb4e2c561e54c3986a0be28cf7))

### Bug Fixes

- handle months that are zero ([1386574](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/138657416d7fb61d2bdf973252f23737357f8c9c))
- use utc time ([24cfb00](https://github.com/lorenzvanherwaarden/datetime-web-component/commit/24cfb007d87b33062af72e5c4d15033342b9ac56))

src/components/style.ts
const css = `
:host {
  --font-family: sans-serif;
  --font-size: 14px;
  --background: #FAFAFA;
  --box-shadow: 0px 4px 16px rgba(0, 0, 0, 0.08);
  --text-color: #222222;
  --muted-text-color: #777777;
  --disabled-text-color: #AAAAAA;
  --focus-color: #0066FF;
  --cell-width: 28px;
  --cell-height: 28px;
  --cell-border-radius: 6px;
  --cell-vertical-margin: 2px;
  --cell-horizontal-margin: 2px;
  --selected-cell-background: #E7E7E7;
  --selected-cell-font-weight: 700;
  --hover-cell-background: #EFEFEF;
  --border-color: #DDDDDD;
  --border-width: 1px;
  --border-radius: 4px;
  --padding: 16px;
  --input-font-size: 18px;
  --input-font-weight: 500;
  --input-border-radius: 4px;
  --input-hover-background: #EEEEEE;
}

@media (prefers-color-scheme: dark) {
  :host {
    --background: #333333;
    --box-shadow: 0px 4px 16px rgba(20, 20, 20, 0.2);
    --text-color: #FFFFFF;
    --muted-text-color: #BBBBBB;
    --disabled-text-color: #777777;
    --focus-color: #4499FF;
    --selected-cell-background: #555555;
    --hover-cell-background: #444444;
    --border-color: #444444;
    --input-hover-background: #444444;
  }
}

:host .dropdown {
  display: block;
  position: absolute;
  z-index: 100000;
  width: calc(7 * var(--cell-width) + 2 * 7 * var(--cell-horizontal-margin));
  font-family: var(--font-family);
  font-size: var(--font-size);
  color: var(--text-color);
  background-color: var(--background);
  border: solid var(--border-width) var(--border-color);
  border-radius: var(--border-radius);
  box-shadow: var(--box-shadow); 
  padding: var(--padding);
}

:host([is-hidden]) .dropdown {
  display: none;
}

.cell {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: auto;
  display: inline-block;
  width: var(--cell-width);
  height: var(--cell-height);
  line-height: var(--cell-height);
  margin: var(--cell-vertical-margin) var(--cell-horizontal-margin);
  vertical-align: middle;
  text-align: center;
  user-select: none;
  border-radius: var(--cell-border-radius);
  outline-color: var(--focus-color);
}

.cell--default {
  cursor: pointer;
}

.cell--default:hover {
  background-color: var(--hover-cell-background);
}

.cell--default:active {
  color: var(--focus-color);
}

.cell--blocked {
  color: var(--disabled-text-color);
  cursor: not-allowed;
}

.cell--selected {
  background-color: var(--selected-cell-background);
  font-weight: var(--selected-cell-font-weight);
}

.cell--header {
  font-weight: 600;
}

.cell--month {
  width: calc(100% / 2 - 2 * var(--cell-horizontal-margin));
}

.header {
  display: flex;
  align-items: center;
  justify-content: space-between;
  user-select: none;
  margin-bottom: 6px;
}

.semantic-button {
  font-family: inherit;
  font-size: var(--font-size);
  background: none;
  color: var(--muted-text-color);
  border: none;
  cursor: pointer;
  padding: 4px;
  outline-color: var(--focus-color);
}

.semantic-button:hover {
  color: var(--focus-color);
}

.semantic-button--icon {
  font-family: system-ui;
  font-size: calc(var(--font-size) + 4px);
  padding: 4px 8px;
}

.semantic-button--icon {
  font-family: system-ui;
}

.container {
  display: flex;
  align-items: center;
  justify-content: center;
}

.container--time {
  margin-top: 10px;
}

.container--year {
  margin: 30px;
}

.input {
  color: var(--text-color);
  font-weight: var(--input-font-weight);
  font-size: var(--input-font-size);
  font-family: var(--font-family);
  background: transparent;
  border: none;
  border-radius: var(--input-border-radius);
  outline-color: var(--focus-color);
  line-height: 1;
  padding: 2px 0px 2px 2px;
  width: auto;
  max-width: 4ch;
}

.input:hover {
  background: var(--input-hover-background);
}

.input--year {
  max-width: 6ch;
}

.time-separator {
  font-size: var(--input-font-size);
  font-weight: var(--input-font-weight);
  margin: 0 4px;
}
`

const style = document.createElement('style')
style.textContent = css

export default style
